contracts/CandidateHub.sol

```diff
@@ -96,28 +118,47 @@ contract CandidateHub is ICandidateHub, System, IParamSubscriber {
     alreadyInit = true;
   }
   
+  function setControlRoundTimeTag(bool value) external {
+    controlRoundTimeTag = value;
+  }
+  
   /********************* ICandidateHub interface ****************************/
   /// Whether users can delegate on a validator candidate
   /// @param agent The operator address of the validator candidate
   /// @return true/false
   function canDelegate(address agent) external override view returns(bool) {
-    uint256 index = operateMap[agent];
-    if (index == 0) {
+    uint256 indexPlus1 = operateMap[agent];
+    if (indexPlus1 == 0) {
       return false;
     }
-    uint256 status = candidateSet[index - 1].status;
+    uint index_ = indexPlus1 - 1;
+    uint256 status = candidateSet[index_].status;
     return status == (status & ACTIVE_STATUS);
   }
 
-  /// Jail a validator for some rounds and slash some amount of deposits
-  /// @param operateAddress The operator address of the validator
-  /// @param round The number of rounds to jail
-  /// @param fine The amount of deposits to slash
-  function jailValidator(address operateAddress, uint256 round, uint256 fine) external override onlyValidator {
-    uint256 index = operateMap[operateAddress];
-    if (index == 0) return;
 
-    Candidate storage c = candidateSet[index - 1];
+/* @product Jail a validator for some rounds and slash some amount of deposits
+   @param operateAddress: The operator address of the validator
+   @param round: The number of rounds to jail
+   @param fine: The amount of deposits to slash
+   @logic
+      1. if the candidate's margin is greater or equal to the sum of the candidate's fine plus
+         the global dues
+          a. set the release round of the candidate to be the current round plus the
+             'round' parameter (if the candidate has prior jail period - add to it
+             the current 'round' parameter)
+          b. subtract the fine's value from the candidate's margin
+          c. and transfer the fine value to the SystemReward contarct
+
+      2. Else:
+          a. remove the candidate from internal structures, and
+          b. transfer the candidate's margin eth value to the SystemReward contract
+  */
+  function jailValidator(address operateAddress, uint256 round, uint256 fine)
+        external override onlyValidator onlyIfCandidateAddress(operateAddress) {
```

Comment:
Here, you make the return logic into modifier.
This method is invoked by ValidatorSet.felony <- SlashIndicator
If the modifier check failed. The whole logic may recover including slash validator's income in ValidatorSet.felony.
That is not our expected.

Reply:
does your comment about return-rather-than-revert logic applies for both onlyValidator and onlyIfCandidateAddress modifiers?